homeassistant/components/imap/coordinator.py

```diff
@@ -0,0 +1,112 @@
+"""Coordinator for imag integration."""
+from __future__ import annotations
+
+import asyncio
+import logging
+from typing import Any
+
+from aioimaplib import AUTH, IMAP4_SSL, SELECTED, AioImapException
+import async_timeout
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_PORT, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import CONF_CHARSET, CONF_FOLDER, CONF_SEARCH, CONF_SERVER, DOMAIN
+from .errors import InvalidAuth, InvalidFolder
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def connect_to_server(data: dict[str, Any]) -> IMAP4_SSL:
+    """Connect to imap server and return client."""
+    client = IMAP4_SSL(data[CONF_SERVER], data[CONF_PORT])
+    await client.wait_hello_from_server()
+    await client.login(data[CONF_USERNAME], data[CONF_PASSWORD])
+    if client.protocol.state != AUTH:
+        raise InvalidAuth
+    await client.select(data[CONF_FOLDER])
+    if client.protocol.state != SELECTED:
+        raise InvalidFolder
+    return client
+
+
+class ImapDataUpdateCoordinator(DataUpdateCoordinator[int]):
+    """Class for imap client."""
+
+    config_entry: ConfigEntry
+
+    def __init__(self, hass: HomeAssistant, imap_client: IMAP4_SSL) -> None:
+        """Initiate imap client."""
+        self.hass = hass
+        self.imap_client = imap_client
+        self.idle_loop_task: asyncio.Task | None = None
+        self.support_push = imap_client.has_capability("IDLE")
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+        )
+
+    async def _async_update_data(self) -> int:
+        """Update the number of unread emails."""
+        try:
+            if self.imap_client is None:
+                await self.retry_connection()
+        except (AioImapException, asyncio.TimeoutError) as err:
+            raise UpdateFailed(err) from err
+
+        return await self.refresh_email_count()
+
+    async def refresh_email_count(self) -> int:
+        """Check the number of found emails."""
+        try:
+            await self.imap_client.noop()
+            result, lines = await self.imap_client.search(
+                self.config_entry.data[CONF_SEARCH],
+                charset=self.config_entry.data[CONF_CHARSET],
+            )
+        except (AioImapException, asyncio.TimeoutError) as err:
+            raise UpdateFailed(err) from err
```

Comment:
@MartinHjelmare , I am testing with different mail servers and noticed that with outlook I am getting a timeout error here sometimes. In such case the entity becomes unavailable and will only update when there is a new push from the server.
Can we schedule a refresh request here before raising UpdateFailed?

Reply:
Yeah, but couldn't that lead to a polling loop?

Reply:
After some testing I found that the timeout error happens when a coordinator update gets called while the imap_client has a pending idle state. For some unknown reason the `self.async_request_refresh` is causing the coordinator to update a second time after being called initially from this line:
https://github.com/home-assistant/core/blob/16295d2d29e3acde87fcd2e4f23cb2f82e2d90e3/homeassistant/components/imap/coordinator.py#L88
Changing this to `self.async_refresh` seems to solve this issue so we only try to update once for each server push.
The other issue that showed up during this testing is that if a user request a manual update using the  `update_entity` service, the update will timeout as the imap_client is in an idle state waiting for a server push. For this case I am not sure if it is possible to disable the `update_entity` service for this integration (is that possible?), otherwise, we have to cancel the idle_loop, run the update and then create a new task for the idle_loop.

Reply:
`async_request_refresh` will use the debouncer while `async_refresh` will not use the debouncer. The default debounce time is 10 seconds. If another refresh was already scheduled within the debounce time the debouncer will schedule the refresh after the debounce time. This means the refresh may be done when the client is in the idle state again when using `async_request_refresh`, as you noticed.

We can overwrite the `async_update` method of the coordinator entity and manage the refresh flow as needed to allow the update entity service to work.

Reply:
I took another approach in commit (https://github.com/home-assistant/core/pull/74623/commits/a03948231b37e7ef08e827d09713c59ec555d90c) regarding the entity update and waiting for a server push.
Instead of having an idle loop, the coordinator will update the entity and then put the imap_client into idle mode until a push is received. Once a push is received the coordinator will request an update and after that is completed it will go into idle mode again (to avoid collision between idling and updating).